Content.Goobstation.Client/Factory/UI/ConstructorBUI.cs

```diff
@@ -0,0 +1,204 @@
+
+        // god BLESS whoever made construction ui for having it so decoupled <3
+        _menu = this.CreateWindow<ConstructionMenu>();
+        PopulateCategories();
+        PopulateRecipes(string.Empty, string.Empty);
+        _menu.PopulateRecipes += (_, args) => PopulateRecipes(args.Item1, args.Item2);
+        _menu.RecipeSelected += (_, item) =>
+        {
+            _menu.ClearRecipeInfo();
+            if (item?.Metadata is ConstructionPrototype proto)
+            {
+                _id = proto.ID;
+                _menu.SetRecipeInfo(proto.Name, proto.Description, _sprite.Frame0(proto.Icon),
+                    proto.Type != ConstructionType.Item, true); // TODO: favourites
+
+                GenerateStepList(proto);
+            }
+            else
+            {
+                _id = null;
+            }
+        };
+        _menu.BuildButtonToggled += (_, _) =>
+        {
+            SendPredictedMessage(new ConstructorSetProtoMessage(_id));
+            _menu.Close();
+        };
+    }
+
+    private void PopulateCategories(string? selected = null)
+    {
+        if (_menu is not {} menu)
+            return;
+
+        var categories = new HashSet<string>();
+
+        foreach (var prototype in _proto.EnumeratePrototypes<ConstructionPrototype>())
+        {
+            var category = prototype.Category;
+
+            if (!string.IsNullOrEmpty(category))
+                categories.Add(category);
+        }
+
+        var categoriesArray = new string[categories.Count + 1];
+
+        // hard-coded to show all recipes
+        var idx = 0;
+        categoriesArray[idx++] = _forAllCategoryName;
+
+        foreach (var cat in categories.OrderBy(Loc.GetString))
+        {
+            categoriesArray[idx++] = cat;
+        }
+
+        menu.OptionCategories.Clear();
+
+        for (var i = 0; i < categoriesArray.Length; i++)
+        {
+            menu.OptionCategories.AddItem(Loc.GetString(categoriesArray[i]), i);
+
+            if (!string.IsNullOrEmpty(selected) && selected == categoriesArray[i])
+                menu.OptionCategories.SelectId(i);
+        }
+
+        menu.Categories = categoriesArray;
+    }
+
+    // copypasted and optimised from ConstructionMenuPresenter
+    private void PopulateRecipes(string search, string category)
+    {
+        if (PlayerManager.LocalEntity is not {} user || _menu is not {} menu)
+            return;
+
+        search = search.Trim().ToLowerInvariant();
+        var searching = !string.IsNullOrEmpty(search);
+        var isEmptyCategory = string.IsNullOrEmpty(category) || category == _forAllCategoryName;
+
+        _recipes.Clear();
+        foreach (var recipe in _proto.EnumeratePrototypes<ConstructionPrototype>())
+        {
+            if (recipe.Hide)
+                continue;
+
+            if (_whitelist.IsWhitelistFail(recipe.EntityWhitelist, user))
+                continue;
+
+            if (searching && !recipe.Name.ToLowerInvariant().Contains(search))
```

Comment:
Normally I'd merge these together into one really ugly if but i'll let this one slide.
```suggestion
            if (searching 
                && !recipe.Name.ToLowerInvariant().Contains(search))